### ISSUE TITLE
3/20 + Test provider permissions (Prevent or at least handle LSV timeouts) + 21

### ISSUE THREAD

**Issue by holbrooa** [lzim/teampsd#429]
**Describe the opportunity**
LSV query timeout requirements are strict, so large teams are pretty likely to get a timeout error, depending on what time of day the query is run. Also, non-provider facilitators should run queries against CDWWork instead. @holbrooa will need to build in some extra database logic to try to prevent this, or at least trap and explain the error and workaround.

**1. Please describe the steps to reproduce the behavior.**
Try to get team data for an entire facility (all clinics)

**2. Please describe the expected behavior.**
Queries should return results if possible, even if it takes a while. Back end should somehow switch between databases dynamically.

**STOP HERE. INFORMATION BELOW IS TO BE FILLED OUT BY WORKGROUPS AS PART OF TRIAGE AND WORK BREAKDOWN PROCESS**

**Estimated person hours to complete:**
21 hours

**Key people (use @ assignment for people whose input will be necessary):**
none
**Key dependencies:**
Are there other tasks that need to be completed first?
none
**Constraints:**
Is there a resource constraint that may affect the development of the feature?
lsv timeout logic (5 mins in the morning, 15 otherwise. Per batch I think, but I'll check the specs to see if it's per query)

This needs to be done before propagation (obviously).

**Gantt chart needed? Attach link and/or screenshot.**
No gantt needed

**Comment by staceypark** [comment]
@holbrooa where are we on this?

**Comment by holbrooa** [comment]
I think what I've implemented here is working, but parts of it I can't test myself. So it's been hanging out in testing. I will need someone else to help with getting this ticket closed, but I'm not sure exactly who I'll need.

**Comment by staceypark** [comment]
@lzim will give all facilitators same access to data UI that she has.
@holbrooa needs to test with persons with only provider access (@lzim will work with Dexter to test this).

Time outs will become inevitable so @holbrooa will need to write an updated message to the end user as LSV requirements become more strict.

**Comment by lzim** [comment]
@lzim 

**Cross ref #429**

- [ ] Add _MTL_ Co-Facilitators to the Operations Access Permissions list for this CDW workgroup
- [ ] Test provider permissions with residents on Tuesday 9/10.

**Comment by staceypark** [comment]
@lzim @holbrooa I've updated the issue title to be reflective of the last piece hanging out about this issue, to help us when triaging in the tracker. Feel free to change it back if I've made it more confusing.

**Comment by staceypark** [comment]
@holbrooa I tested out the data UI buttons with a clinician and both buttons give a permissions error. I took screenshots of the error pop up and of the window that pops up when you hit debug. Sent to you in email.

**Comment by staceypark** [comment]
@lzim 
Given that the provider test has failed, @holbrooa says that his workaround plan did not work. He recommends that @lzim give permissions to clinicians specifically. Since it will not be feasible to go through the permissions process for all clinicians for all sites, @holbrooa recommends that we give permissions to point persons (i.e. team leads) on each team.

**Comment by lzim** [comment]
@holbrooa and @staceypark 

So, the functionality during the R21 - all frontline providers could query - is now problematic/unresolved.

Questions:
1. Is this for all servers or just those that use LSV? 
2. What proportion (nationally) will be using LSV?
3. What is meant by LSV becoming more strict?
4. How does this relate to the timeouts?
5. Can Andrew provide any other leads or contacts on granting permissions?

Andrew's Recommendation is to provide - Operations Access to Team Leads, but this does remove the accessibility principle we've sought since the beginning.

**Comment by holbrooa** [comment]
We've been forced to change servers a couple of times. At this point, we are locked in to the field reporting enclave (FRE). It does have LSV, but my understanding is that the new lsv timeout thresholds are too strict for us. Basically, they automatically kill any query that takes more than 5 to 15 minutes, depending on the time of day. I have a couple of ideas for how we could maybe test some workarounds, but I can't try it myself, so I've been left with  "coding blind" so to speak.  There's also the problem of having implemented some of our own data rules as tables we need to join to (e.g. Jodie's PC ICD codes) which mean I've had to join to our own tables in PTSD_OMHO. Obviously providers don't automatically have access to that. I can rework the queries to not require that join, at the expense of making them more complicated and take a bit longer. That trade-off isn't a no-brainer though.

So what we've been doing is using the regular CDW instead of LSV, to make sure we don't get auto-killed.  It really feels like what we're trying to do isn't consistent with how the CDW folks envision any data resources being used. Hence the repeated switching from server to server as they bring the resources that fit our purpose offline. I'll try to build a few different options or at least ideas for how to handle this in the data UI code, to be used when there is an opportunity to do more testing with a provider.

**Comment by staceypark** [comment]
@holbrooa please add link to where you can check other folks' permissions.
@lzim 
LSV permissions grant backend access to things you already have frontend access to. 
1. LSV query has stricter permissions than national operations access for example. @holbrooa has tried to keep us below the timeout threshold, but could still run into timeout, especially at certain times of the day.
2. As long as we have instructions for facilitators on requesting LSV access within VISTA, we can teach learners how to request access.
3. @holbrooa will attempt to set up so that whoever does the propagation of the data UI, which sets the data UI to a certain station and populates the clinic list, it will perform the lookup of PTSD_OMHO at that point which means that person has to have healthcare operations permissions and it means that the learner cannot change the station. If the learner really needs to look at different station, they can navigate to the Splashpage and select the data UI for that station.

Note: VISTA is phasing out @lzim @staceypark follow up on the VISTA phasing out schedule.
@staceypark need to think through training for checking & updating LSV perms.

**Comment by ritahitching** [comment]
@holbrooa Do you have an update on where we are at with:

1. co-facilitator & test provider permissions 

1. commensurate documentation

**Comment by holbrooa** [comment]
I believe I'm good with the new system utilizing local lookups that get updated when the data UI itself gets mapped to a station. That was a decent-sized change to how the thing works. It should now correctly use lsv perms (and only that) which you can test by pointing the provider you had try it before to the vista - network username thing linked above. If he does that step and waits for the system to reset, he should then be able to use the data UI's buttons for data local to his station. 

I am now in permissions purgatory, so I can't log in and show you these things directly anymore. At least, not until I'm back up and running.

**Comment by holbrooa** [comment]
Oh, the thing I didn't get to is making a version that doesn't use lsv at all and just uses operations access. That's a very simple change that I could talk someone through, though.  Just a bit of find-and-replace.

**Comment by staceypark** [comment]
@lzim I believe this has been tested and resolved as of the Monday Jan 6 QIIC meeting?

**Comment by staceypark** [comment]
ISSUE: Last time we tested this with M. Freeman CPRS permissions and it did not work.
We need to get to the bottom of this ASAP, with the launch coming up soon and the memo including mtl.how/data rolling out very soon

**Comment by staceypark** [comment]
**Summary of discussion**
CDWWork
- NDS = Here is who we ask for Operations permissions.
- VINCI = Here is who we ask for Research permissions.
- VIREC
- BISL = Of all the types of CDW SharePoints, we use the BISL one specifically.  

1. We chose the BISL SharePoint because this one was supposed to help us with access for frontline staff permissions.
2. Then, the VA switched to LSV.
3. Now we are encountering permissions issues for frontline staff again at Nevermore FRE14

@anthonycpichardo @staceypark @anazariz 
We could check our own permissions at:
CDW Home > LSV > Check permissions

**Comment by anazariz** [comment]
@lzim , @anthonycpichardo 
If OK with you, I need to move this portion of the ticket to Done.  We may need to create a new ticket for the remediation part - i.e. how to get access for those who  don't have it.  
From the perspective of creating a technical solution, that piece is done.

**Comment by anazariz** [comment]
Each VISN connects to a unique server containing an LSV database).There needs to be a change in the data workbook (not names here) that connects to its corresponding database server.

**Comment by lzim** [comment]
> Each VISN connects to a unique server containing an LSV database).There needs to be a change in the data workbook (not names here) that connects to its corresponding database server.

@lzim @staceypark @anthonycpichardo @anazariz 
1. The goal of issue card #429 is that Team PSD does not have to manage permissions, rather they are managed locally.
2. We will not be granting permissions to the national operations server.
3. Please review the LSV documentation in the CDW documentation that we reviewed in our meeting back in February.
4. As I mentioned to our call, I also have questions about the test that was circulated to the QIICs and whether it actually was a test of LSV permissions. 
5. Please review the email that I sent about this in terms of the goal of having QIICs who each have their own local permissions test the LSV buttons only

**Comment by anthonycpichardo** [comment]
CLosing this and work is continued on mtl_code#6
